### DISCUSSION TITLE
Ability to run script before install

### DISCUSSION THREAD

**Discussion by mlahargou** [renovatebot/renovate#29254] [Suggest an Idea]
Background:
--
We use [syncpack](https://github.com/JamieMason/syncpack) to enforce certain rules about dependencies on our pnpm monorepo. We'd like to be able to run `syncpack fix-mismatches` after the dependencies have been updated but **before** renovate runs `pnpm install`. 

Currently, the only way to run arbitrary scripts is with `postUpgradeTasks` but those run **after** the `pnpm install`. So, we have to run `pnpm install` manually again:

```
      {
         matchManagers: ['npm'],
         postUpgradeTasks: {
            commands: ['pnpm syncpack fix-mismatches && pnpm install'],
            fileFilters: ['**/pnpm-lock.yaml', '**/package.json'],
            executionMode: 'branch',
         },
      },
```

Basically, this is the current flow:

`upgrade deps` -> `pnpm install` -> (`pnpm syncpack fix-mismatches` -> `pnpm install`)

but we would rather be able to:

 `upgrade deps` -> `pnpm syncpack fix-mismatches` -> `pnpm install`

Possible solution:
--
Have a `preInstallTasks` option (very similar to `postUpgradeTasks`) which just runs arbitrary scripts after the package.jsons have been updated but before the install happens.

Relevant Logs:
--
```
DEBUG: Writing any updated package files (repository=..., branch=renovate-update-react-monorepo)
...
DEBUG: Executing command (repository=..., branch= renovate-update-react-monorepo)
       "command": "install-tool node 18.19.1"
...
DEBUG: Executing command (repository=..., branch= renovate-update-react-monorepo)
       "command": "install-tool pnpm 9.0.6"
...
DEBUG: Executing command (repository=..., branch= renovate-update-react-monorepo)
       "command": "pnpm install --recursive --lockfile-only"
...
DEBUG: Executing command (repository=..., branch=renovate-update-react-monorepo)
       "dep": "@types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react @types/react-dom @types/react-dom @types/react-dom @types/react-dom @types/react-dom @types/react-dom react react react react react react react react react react react react react react react react react react react react react react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-dom react-is react-is react-is",
       "command": "pnpm syncpack fix-mismatches && pnpm install"

**Comment by mlahargou** [comment]
Another, albeit hacky, solution would be to use a combination of pnpm's `pnpm:devPreinstall` script and `customEnvVariables`.  (I haven't tested this yet)

So, adding something like `"pnpm:devPreinstall": "[ -z $RENOVATE ] || pnpm syncpack fix-mismatches"` to the root package.json. Then set a `RENOVATE=true` variable with [`customEnvVariables`](https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables).

**Reply by mlahargou** [reply]
I tried this out and it doesn't work because the `pnpm:devPreinstall` doesn't run when the `--lockfile-only` arg is added to `pnpm install`.

**Comment by svimercati** [comment]
@mlahargou just a quick question, I'm a newbie with this tool, could you share me how did you enable renovate to run pnpm install, because using the base configuration

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended",
    "docker:pinDigests",
    "npm:unpublishSafe",
    "packages:jsTest",
    "packages:jsUnitTest",
    "packages:linters",
    "packages:postcss",
    "packages:react",
    "packages:tslint",
    "packages:vite",
    "customManagers:dockerfileVersions",
    "replacements:all",
    ":docker",
    ":npm"
  ]
}
```

what it's happening is that PR are raised but pnpm-lock file is not updated and so when the pipeline run it fails because there is no matching with package.json libraries that renovate has updated. One of the options that I was thinking on was to run it in postUpgradeTasks as a command but from what I read from you this is something configured in another way?

Regards

**Reply by mlahargou** [reply]
If I understand your question, you are having trouble getting pnpm to work with renovate? That seems like a separate issue, so I'll be brief. As long as you have added `npm` to the `enabledManagers`, it should update the `pnpm-lock.yaml` for you (as there is [support for updating the pnpm lock file](https://github.com/renovatebot/renovate/blob/5d99ff9862bac803dcea4e6ded5f3fec34abccfc/lib/manager/npm/post-update/pnpm.ts)). We also force the pnpm version and run `pnpm dedupe` in `postUpdateOptions`. If you are having issues past that, I think you should look at the debug logs of renovate and try to see where the error is occurring.

```
   enabledManagers: ['npm'],
   force: {
      constraints: {
         pnpm: '9.0.6',
      },
   },
   postUpdateOptions: ['pnpmDedupe'],
```

The case I'm talking about is a very special case where I want to run another script **before** the `pnpm install` happens.

**Comment by mlahargou** [comment]
Closing as this seems like a duplicate of https://github.com/renovatebot/renovate/issues/8804
